File: ru.kmorozov.library.data.model/src/main/java/ru/kmorozov/library/data/storage/mongo/LikeTextSearch.java
package ru.kmorozov.library.data.storage.mongo;

import com.mongodb.BasicDBObject;
import org.bson.Document;
import org.bson.types.ObjectId;
import org.springframework.data.mongodb.core.MongoTemplate;

import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;

/**
 * Created by sbt-morozov-kv on 29.11.2016.
 */
public class LikeTextSearch {

    private static final int SEARCH_LIMIT = 100;

    private final String collectionName;
    private final MongoTemplate mongoTemplate;

    public LikeTextSearch(final String collectionName, final MongoTemplate mongoTemplate) {
        this.collectionName = collectionName;
        this.mongoTemplate = mongoTemplate;
    }

    public Collection<ObjectId> findMatchingIds(final String searchString) {
        final Document result = executeFullTextSearch(searchString);
        return extractSearchResultIds(result);
    }

    private Document executeFullTextSearch(final String searchString) {
        final Document textSearch = new Document();
        textSearch.put("$text", collectionName);
        textSearch.put("search", searchString);
        textSearch.put("limit", SEARCH_LIMIT); // override default of 100
        textSearch.put("project", new Document("_id", 1));
        return mongoTemplate.executeCommand(textSearch);
    }

    private static Collection<ObjectId> extractSearchResultIds(final Document commandResult) {
        final Collection<ObjectId> objectIds = new HashSet<>();
        final Iterable resultList = (Iterable) commandResult.get("results");

        if (null == resultList) return Collections.emptyList();

        for (final Object aResultList : resultList) {
            final BasicDBObject resultContainer = (BasicDBObject) aResultList;
            final BasicDBObject resultObj = (BasicDBObject) resultContainer.get("obj");
            final ObjectId resultId = (ObjectId) resultObj.get("_id");
            objectIds.add(resultId);
        }
        return objectIds;
    }

}

File: ru.kmorozov.gbd.db.mongo/src/main/java/ru/kmorozov/library/data/storage/mongo/LikeTextSearch.kt
package ru.kmorozov.library.data.storage.mongo

import com.mongodb.BasicDBObject
import org.bson.Document
import org.bson.types.ObjectId
import org.springframework.data.mongodb.core.MongoTemplate
import java.util.HashSet

/**
 * Created by sbt-morozov-kv on 29.11.2016.
 */
class LikeTextSearch(private val collectionName: String, private val mongoTemplate: MongoTemplate) {

    fun findMatchingIds(searchString: String): Collection<ObjectId> {
        val result = executeFullTextSearch(searchString)
        return extractSearchResultIds(result)
    }

    private fun executeFullTextSearch(searchString: String): Document {
        val textSearch = Document()
        textSearch["\$text"] = collectionName
        textSearch["search"] = searchString
        textSearch["limit"] = SEARCH_LIMIT // override default of 100
        textSearch["project"] = Document("_id", 1)
        return mongoTemplate.executeCommand(textSearch)
    }

    companion object {

        private const val SEARCH_LIMIT = 100

        private fun extractSearchResultIds(commandResult: Document): Collection<ObjectId> {
            val objectIds = HashSet<ObjectId>()
            val resultList = commandResult["results"] as Iterable<*>

            for (aResultList in resultList) {
                val resultContainer = aResultList as BasicDBObject
                val resultObj = resultContainer.get("obj") as BasicDBObject
                val resultId = resultObj.get("_id") as ObjectId
                objectIds.add(resultId)
            }
            return objectIds
        }
    }

}
